### PR TITLE
Adding GetSchemaTable() on SqlDataReader

### PR DESCRIPTION
**Pull request by saurabh500**
Porting over code to add GetSchemaTable in .Net Core SqlClient. 
The code is a port from NetFx code. The `CheckSparseColumnBit` test is a port from NetFx, 

Reduces failures in Dapper tests by 5 based on my internal runs. 

I ran the EF functional tests with my changes. No regressions observed with the changes. 

Fixes https://github.com/dotnet/corefx/issues/19748

cc @stephentoub @corivera @danmosemsft @karelz

### CODE REVIEW

**Comment by karelz** [comment]
Sounds reasonable to me, given it is driven by direct customer feedback - @danmosemsft @Petermarcu are you ok with taking it?

cc @NickCraver

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 1]
```diff
@@ -395,6 +396,266 @@ internal void Bind(TdsParserStateObject stateObj)
...
+            DataColumn IsExpression = new DataColumn(SchemaTableColumn.IsExpression, typeof(System.Boolean));
+            DataColumn IsIdentity = new DataColumn("IsIdentity", typeof(System.Boolean));
+            DataColumn DataTypeName = new DataColumn("DataTypeName", typeof(System.String));
+            DataColumn UdtAssemblyQualifiedName = new DataColumn("UdtAssemblyQualifiedName", typeof(System.String));
+            // Xml metadata specific
+            DataColumn XmlSchemaCollectionDatabase = new DataColumn("XmlSchemaCollectionDatabase", typeof(System.String));
+            DataColumn XmlSchemaCollectionOwningSchema = new DataColumn("XmlSchemaCollectionOwningSchema", typeof(System.String));
+            DataColumn XmlSchemaCollectionName = new DataColumn("XmlSchemaCollectionName", typeof(System.String));
+            // SparseColumnSet
+            DataColumn IsColumnSet = new DataColumn("IsColumnSet", typeof(System.Boolean));
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Nit: it'd be nice if all of these locals were camel cased

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 2]
```diff
@@ -1063,6 +1324,32 @@ override public int GetProviderSpecificValues(object[] values)
...
+        {
+            SqlStatistics statistics = null;
+            try
+            {
+                statistics = SqlStatistics.StartTimer(Statistics);
+                if (null == _metaData || null == _metaData.schemaTable)
+                {
+                    if (null != this.MetaData)
+                    {
+
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Nit: unnecessary blank line

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 3]
```diff
@@ -1063,6 +1324,32 @@ override public int GetProviderSpecificValues(object[] values)
...
+                        _metaData.schemaTable = BuildSchemaTable();
+                        Debug.Assert(null != _metaData.schemaTable, "No schema information yet!");
+                    }
+                }
+                if (null != _metaData)
+                {
+                    return _metaData.schemaTable;
+                }
+                return null;
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Nit: could just be `return _metadata?.schemaTable;`

File: src/System.Data.SqlClient/tests/ManualTests/SQL/DataReaderTest/DataReaderTest.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+using Xunit;
+
+namespace System.Data.SqlClient.ManualTesting.Tests
+{
+    public static class DataReaderTest
+    {
+        [CheckConnStrSetupFact]
+        public static void LoadReaderIntoDataTableToTestGetSchemaTable()
+        {
+            using(SqlConnection connection = new SqlConnection(DataTestUtility.TcpConnStr)) {
```
Comment (stephentoub) [inline src/System.Data.SqlClient/tests/ManualTests/SQL/DataReaderTest/DataReaderTest.cs]: Nit: brace on next line

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs [hunk 5]
```diff
@@ -1063,6 +1324,32 @@ override public int GetProviderSpecificValues(object[] values)
             return GetSqlValues(values);
         }
 
+        override public DataTable GetSchemaTable()
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]: Nit: public override

**Comment by saurabh500** [comment]
@karelz Should I merge this or wait for an ack from Peter or Dan?

**Comment by danmosemsft** [comment]
We do not need Peter's ack for a change of this sort. This is not a DCR. 
Also happily the fork has still not happened yet.

**Pull request merged**